### PR TITLE
Added disk space warning if not enough for ES reindex

### PR DESCRIPTION
**Pull request by jashaik**
Signed-off-by: jan shahid shaik <jashaik@progress.com>

### Description

Raise disk space warning if disk space is not enough for ES reindexing

### Issues Resolved

Added function to verify disk space in the preflight checks

https://app.zenhub.com/workspaces/chef-infra-server-team-5fc64867d45ca500173dbbc7/issues/chef/chef-server/2197

### Check List

- [ ] New functionality includes tests
- [ ] All buildkite tests pass
- [ ] Full omnibus build and tests in buildkite pass
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [ ] PR title is a worthy inclusion in the CHANGELOG

### CODE REVIEW

**Comment by jashaik** [comment]
Pipeline information
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2279
Umbrella - https://buildkite.com/chef/chef-umbrella-master-chef-server/builds/300

File: src/chef-server-ctl/plugins/du.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+require 'mixlib/shellout'
```
Comment (marcparadise) [inline src/chef-server-ctl/plugins/du.rb:1]: This file (du.rb) should live in lib/chef_server_ctl/du.rb, under `module ChefServerCtl, because it is not a plugin.

File: src/chef-server-ctl/plugins/reindex.rb [hunk 2]
```diff
@@ -63,6 +63,27 @@ def do_reindex(orgs_to_reindex, options)
...
+      puts "Free space is sufficient to start Elasticsearch reindex"
+    else
+      puts "Insufficient free space on disk to complete reindex."
+      puts "The current elasticsearch data directory contains #{data_dir_size} KB of data but only #{free_disk_space} KB is available on disk."
+      puts "The reindex process requires at least #{2*data_dir_size} KB."
+      exit 1
+    end
+  else
+    puts "Elasticserach data path does not exist, so skipping the disk space preflight validations: #{data_dir}"
+  end
```
Comment (marcparadise) [inline src/chef-server-ctl/plugins/reindex.rb:84]: This should end in a `rescue` that shows only the error message/info in case of failure.  That way  if anything in `Du.du` fails it doesn't raise a stack-trace - those are noise and generally unhelpful to the operator. 

To that end, it would be better if `du` failures raised a known type (like a new exception DuFailedException) so that we could rescue it specifically here.

File: src/chef-server-ctl/plugins/statfs.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
+require 'ffi'
+
+class Statfs
```
Comment (marcparadise) [inline src/chef-server-ctl/plugins/statfs.rb:3]: This should also be in lib/chef_server_ctl.

File: src/chef-server-ctl/plugins/statfs.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    # conservative.
+    (@statvfs[:f_frsize] * @statvfs[:f_bavail]) / 1024
+  end
+
+  private
+
+  def stat(path)
+    statvfs = Statvfs.new
+    if statvfs(path, statvfs.to_ptr) != 0
+      raise 'statvfs: ' + strerror(FFI.errno)
```
Comment (marcparadise) [inline src/chef-server-ctl/plugins/statfs.rb:74]: Let's also raise a new named exception type here, so that we can catch it in the caller and handle it better.

**Review by marcparadise — CHANGES_REQUESTED** [review]
Please see comments inline.

**Comment by jashaik** [comment]
New pipeline information
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2291
umbrella - https://buildkite.com/chef/chef-umbrella-master-chef-server/builds/304

File: src/chef-server-ctl/plugins/reindex.rb [hunk 5]
```diff
@@ -63,6 +65,31 @@ def do_reindex(orgs_to_reindex, options)
...
+  begin
+    if Dir.exist?(data_dir)
+      data_dir_size = Du.du(data_dir)
+      free_disk_space = Statfs.new("#{data_dir}/..").free_space
+      if (2.2* data_dir_size) < free_disk_space #The minimum space should be double the existing data size
+        puts "Free space is sufficient to start Elasticsearch reindex"
+      else
+        $stderr.puts "Insufficient free space on disk to complete reindex."
+        $stderr.puts "The current elasticsearch data directory contains #{data_dir_size} KB of data but only #{free_disk_space} KB is available on disk."
+        $stderr.puts "The reindex process requires at least #{2*data_dir_size} KB."
```
Comment (PrajaktaPurohit) [inline src/chef-server-ctl/plugins/reindex.rb:82]: Should this multiplication factor be 2.2 since that is what we are checking on?

File: src/chef-server-ctl/lib/chef_server_ctl/helpers/statfs.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    # conservative.
+    (@statvfs[:f_frsize] * @statvfs[:f_bavail]) / 1024
+  end
+
+  private
+
+  def stat(path)
+    statvfs = Statvfs.new
+    if statvfs(path, statvfs.to_ptr) != 0
+      raise 'StatvfsFailedException'
```
Comment (PrajaktaPurohit) [inline src/chef-server-ctl/lib/chef_server_ctl/helpers/statfs.rb:74]: It might be useful to include the error message here `raise StatvfsFailedException, strerror(FFI.errno)` so that rescue  can `rescue StatvfsFailedException => e` and print e.message.

Similarly for the du error

File: src/chef-server-ctl/plugins/reindex.rb [hunk 7]
```diff
@@ -63,6 +65,31 @@ def do_reindex(orgs_to_reindex, options)
...
+      else
+        $stderr.puts "Insufficient free space on disk to complete reindex."
+        $stderr.puts "The current elasticsearch data directory contains #{data_dir_size} KB of data but only #{free_disk_space} KB is available on disk."
+        $stderr.puts "The reindex process requires at least #{2.2*data_dir_size} KB."
+        exit 1
+      end
+    else
+      puts "Elasticserach data path does not exist, so skipping the disk space verification: #{data_dir}"
+    end
+  rescue => exception
```
Comment (jashaik) [inline src/chef-server-ctl/plugins/reindex.rb:88]: We are giving writing the exception here. Is this what you are trying to suggest me? @PrajaktaPurohit

**Comment by jashaik** [comment]
New Pipeline information - 
Adhoc - https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/2374
Umbrella - https://buildkite.com/chef/chef-umbrella-master-chef-server/builds/332

File: src/chef-server-ctl/plugins/reindex.rb (hunk 7 above)
Comment (PrajaktaPurohit) [reply src/chef-server-ctl/plugins/reindex.rb:88]: yeah - I was thinking it might be of help to catch the individual exception that we are throwing in du and statsvfs instead of rescuing everything.

File: src/chef-server-ctl/lib/chef_server_ctl/helpers/du.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  #
+  def self.du(path)
+    command = Mixlib::ShellOut.new("du -sk #{path}")
+    command.run_command
+    if command.status.success?
+      command.stdout.split("\t").first.to_i
+    else
+      puts "du -sk #{path} failed with exit status: #{command.exitstatus}"
+      puts "du stderr: #{command.stderr}"
+      raise 'DuFailedException', command.stderr(FFI.errno)
```
Comment (marcparadise) [inline src/chef-server-ctl/lib/chef_server_ctl/helpers/du.rb:17]: Does command.stderr accept an argument?

File: src/chef-server-ctl/plugins/reindex.rb (hunk 7 above)
Comment (marcparadise) [reply src/chef-server-ctl/plugins/reindex.rb:88]: While I don't normally like catch-all rescues, I think it makes sense in this context. 

Almost any failure out of these checks would not be actionable by the user, so we'd end up just showing them an error that meant little - and then we'd leave them stuck there with it and no path forward.  (Types of errors I can think of: du command not found, broken permissions, disk failures preventing statfs reads). 

If the error was due to something like hard drive failure or permissions, we'll still fail later - and will still have the logged exception from here to give a hint if one is needed.

TL;DR - I think this should do what it says on the tin (or function name) and only fail if there is not enough disk space. Allowing exceptions to escape would mean that the user ends up with a stack trace and confusion.

File: src/chef-server-ctl/plugins/reindex.rb [hunk 9]
```diff
@@ -63,6 +65,31 @@ def do_reindex(orgs_to_reindex, options)
...
+      if (2.2* data_dir_size) < free_disk_space #The minimum space should be double the existing data size
+        puts "Free space is sufficient to start Elasticsearch reindex"
+      else
+        $stderr.puts "Insufficient free space on disk to complete reindex."
+        $stderr.puts "The current elasticsearch data directory contains #{data_dir_size} KB of data but only #{free_disk_space} KB is available on disk."
+        $stderr.puts "The reindex process requires at least #{2.2*data_dir_size} KB."
+        exit 1
+      end
+    else
+      puts "Elasticserach data path does not exist, so skipping the disk space verification: #{data_dir}"
```
Comment (marcparadise) [inline src/chef-server-ctl/plugins/reindex.rb:86]: Typo in elasticsearch

**Review by marcparadise — CHANGES_REQUESTED** [review]

File: src/chef-server-ctl/lib/chef_server_ctl/helpers/du.rb (hunk 8 above)
Comment (jashaik) [reply src/chef-server-ctl/lib/chef_server_ctl/helpers/du.rb]: Looking like it is not accepting any argument, Do we have any docs for referring command module functions arity.I have verified this by running some smaple tests.

**Review by PrajaktaPurohit — APPROVED** [review]

**Review by marcparadise — APPROVED** [review]
